purchases/src/test/java/com/revenuecat/purchases/attributes/SubscriberAttributesBackendTests.kt

```diff
@@ -27,125 +34,217 @@ class SubscriberAttributesBackendTests {
             appUserID,
-            {
-                fail("should be failure")
-            },
-            { error,  syncedSuccessfully ->
-                receivedError = error
-                receivedSyncedSuccessfully = syncedSuccessfully
-            }
+            unexpectedOnSuccess,
+            expectedOnError
         )
+
         assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.UnknownBackendError)
         assertThat(receivedSyncedSuccessfully).isFalse()
+        assertThat(receivedAttributeErrors).isEmpty()
     }
 
+    // endregion
+
+    // region posting attributes when posting receipt
+
+    private val subscriberAttribute1 = SubscriberAttribute("key", "value")
+    private val subscriberAttribute2 = SubscriberAttribute("key1", null)
+    private val mapOfSubscriberAttributes = mapOf(
+        "key" to subscriberAttribute1,
+        "key1" to subscriberAttribute2
+    )
+    private val fetchToken = "fetch_token"
+    private val productID = "product_id"
+
     @Test
-    fun `attributes validation error when posting attributes`() {
-        mockResponse("/subscribers/$appUserID/attributes", 400, resultBody = "{\n" +
-            "  \"code\": 7262,\n" +
-            "  \"message\": \"Some subscriber attributes keys were unable to saved.\",\n" +
-            "  \"attribute_erors\": [\n" +
-            "    {\n" +
-            "      \"key_name\": \"email\",\n" +
-            "      \"message\": \"Value is not a valid email address.\"\n" +
-            "    }\n" +
-            "  ]\n" +
-            "}")
-
-        var receivedError: PurchasesError? = null
-        var receivedSyncedSuccessfully: Boolean? = null
-        underTest.postSubscriberAttributes(
-            mapOf("email" to SubscriberAttribute("email", null)),
+    fun `posting receipt with attributes works`() {
+        mockPostReceiptResponse(mapOfSubscriberAttributes)
+
+
+        underTest.postReceiptData(
+            fetchToken,
             appUserID,
-            {
-                fail("should be failure")
-            },
-            { error,  syncedSuccessfully ->
-                receivedError = error
-                receivedSyncedSuccessfully = syncedSuccessfully
-            }
+            productID,
+            false,
+            null,
+            false,
+            null,
+            null,
+            mapOfSubscriberAttributes,
+            expectedOnSuccessPurchaserInfo,
+            unexpectedOnError
         )
-        assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.InvalidSubscriberAttributesError)
-        assertThat(receivedSyncedSuccessfully).isTrue()
+
+        assertThat(receivedPurchaserInfo).isNotNull
+        val actualPostReceiptBody = actualPostReceiptBodySlot.captured
+        assertThat(actualPostReceiptBody).isNotNull()
+        assertThat(actualPostReceiptBody["attributes"]).isNotNull
+    }
+
+    @Test
+    fun `posting receipt without attributes skips them`() {
```

Comment:
maybe it's worth adding test cases for when attributes fail to save, to check that the errors are passed along?